### PR TITLE
Add existence check before adding widget

### LINKED ISSUE #8511: Regression when opening an .MD file in editor: I can see two tabs for one editor
See the screencast:

screencast 2020-09-16 16-49-47

- Theia Version: 993b7a8bc

### PR DESCRIPTION
**Pull request by kena0ki**
Signed-off-by: Ken Aoki <ken3574.25@gmail.com>

#### What it does

Fixes #8511 

Before a widget is added, whether widget is attached or not is checked in `widget-open-handler.ts`. 
https://github.com/eclipse-theia/theia/blob/39887bf9d05c29fedc9466db76ca0a58706a1a2a/packages/core/src/browser/widget-open-handler.ts#L98-L100
However a widget is not necessarily attached immediately after the widget is added.
When we double click, click event is fired twice. The first event adds a widget and it can take some time to attach the widget for some reason (like network traffic?). In such a case, the second event can miss the widget exists and unnecessarily adds an additional widget.
This PR adds check for whether a widget exists or not, and prevents an additional widget from being added.

#### How to test

1.  Start browser example using Gitpod. (Maybe any other cloud IDE can be used.)
2. Open a file by double click.
3. Additional empty editor should not be opened.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/core/src/browser/widget-manager.ts [hunk 1]
```diff
@@ -222,6 +218,34 @@ export class WidgetManager {
...
+    }
+
+    /**
+     * Returns whether the wedget exists or not.
+     * @param factoryId The widget factory id.
+     * @param options The widget factory specific information.
+     *
+     * @returns Whether the wedget exists or not.
+     */
+    isWidgetExisting(factoryId: string, options?: any): boolean {
```
Comment (vince-fugnitto) [inline packages/core/src/browser/widget-manager.ts:244]: The naming of the method can be improved.
We can simplify it to `widgetExists` or something along the lines of:

```suggestion
    findWidget(factoryId: string, options?: any): Widget | undefined {
```

If the widget exists, simply return it.

File: packages/git/src/browser/history/git-commit-detail-open-handler.ts [hunk 2]
```diff
@@ -46,10 +46,6 @@ export class GitCommitDetailOpenHandler extends WidgetOpenHandler<GitCommitDetai
         }
     }
 
-    protected async doOpen(widget: GitCommitDetailWidget, options: GitCommitDetailOpenerOptions): Promise<void> {
-        await super.doOpen(widget, options);
-    }
-
```
Comment (vince-fugnitto) [inline packages/git/src/browser/history/git-commit-detail-open-handler.ts:52]: Why is this change necessary?

File: packages/core/src/browser/widget-open-handler.ts [hunk 3]
```diff
@@ -86,16 +86,17 @@ export abstract class WidgetOpenHandler<W extends BaseWidget> implements OpenHan
      * @returns promise of the widget that resolves when the widget has been opened.
      */
     async open(uri: URI, options?: WidgetOpenerOptions): Promise<W> {
-        const widget = await this.getOrCreateWidget(uri, options);
-        await this.doOpen(widget, options);
+        const isWidgetExisting = this.isWidgetExisting(uri, options);
+        const widget = isWidgetExisting ? (await this.getWidget(uri, options))! : await this.createWidget(uri, options);
+        await this.doOpen(widget, isWidgetExisting, options);
```
Comment (vince-fugnitto) [inline packages/core/src/browser/widget-open-handler.ts:91]: I'm not sure I understand the changes, shouldn't we call `getOrCreateWidget` which already checks if the widget exists, and if not creates it?

- https://github.com/eclipse-theia/theia/pull/8913/files#diff-d8c57e795e821606a1336d37fda1835453c6f572e9dc17e99f7ab10e65096fffR228-R235

We are duplicating functionality with this approach.

File: packages/git/src/browser/history/git-commit-detail-open-handler.ts (hunk 2 above)
Comment (kena0ki) [reply packages/git/src/browser/history/git-commit-detail-open-handler.ts:52]: Since number of parameters of `super.doOpen()` is changed, this override method also needs to be changed.
However, I believe this override method has nothing different in behavior from method `super.doOpen()` because it just calling `super.doOpen()`.
So I just removed it.

File: packages/core/src/browser/widget-manager.ts (hunk 1 above)
Comment (kena0ki) [reply packages/core/src/browser/widget-manager.ts:244]: I prefer to choose renaming the method.
Maybe the best way would change `tryGetWidget` to get a widget from `widgetPromises` and `pendingWidgetPromises` not only from `widgetPromises`, if possible.

File: packages/core/src/browser/widget-open-handler.ts (hunk 3 above)
Comment (kena0ki) [reply packages/core/src/browser/widget-open-handler.ts:91]: We could use `getOrCreateWidget` at line 90.
But if we use `getOrCreateWidget`, the check whether a widget exists or not is executed twice.
I think the check result would be the same, but I didn't want to separate the check execution. (If there were "await" between the checks the result could change.)
I don't think I want to persist my solution,  so feel free to decline my opinion.

File: packages/core/src/browser/widget-manager.ts [hunk 4]
```diff
@@ -222,6 +218,34 @@ export class WidgetManager {
...
+        }
+        return this.createWidget(factoryId, options);
+    }
+
+    /**
+     * Returns whether the wedget exists or not.
+     * @param factoryId The widget factory id.
+     * @param options The widget factory specific information.
+     *
+     * @returns Whether the wedget exists or not.
```
[the remark this answers is not in the document]
Comment (kena0ki) [reply packages/core/src/browser/widget-manager.ts:242]: Thank you. I'll fix it.

File: packages/core/src/browser/widget-manager.ts [hunk 5]
```diff
@@ -222,6 +218,34 @@ export class WidgetManager {
...
+        const key = this.toKey({ factoryId, options });
+        const existingWidget = this.doGetWidget<T>(key);
+        if (existingWidget) {
+            return existingWidget;
+        }
+        return this.createWidget(factoryId, options);
+    }
+
+    /**
+     * Returns whether the wedget exists or not.
```
[the remark this answers is not in the document]
Comment (kena0ki) [reply packages/core/src/browser/widget-manager.ts:238]: Thank you. I'll fix it.

[the remark this answers is not in the document]
Comment (kena0ki) [reply packages/core/src/browser/widget-manager.ts]: Done.

File: packages/core/src/browser/widget-manager.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (kena0ki) [reply packages/core/src/browser/widget-manager.ts]: Done.

File: packages/core/src/browser/widget-manager.ts (hunk 1 above)
Comment (kena0ki) [reply packages/core/src/browser/widget-manager.ts]: Changed the name.

**Comment by kena0ki** [comment]
@vince-fugnitto Thank you for the review. I've update the changes.

File: packages/core/src/browser/widget-manager.ts [hunk 6]
```diff
@@ -222,6 +218,34 @@ export class WidgetManager {
...
+    }
+
+    /**
+     * Returns whether the widget exists or not.
+     * @param factoryId The widget factory id.
+     * @param options The widget factory specific information.
+     *
+     * @returns Whether the widget exists or not.
+     */
+    widgetExists(factoryId: string, options?: any): boolean {
```
Comment (colin-grant-work) [inline packages/core/src/browser/widget-manager.ts:244]: I think this is fine, but there are two alternatives:
keep the `widgetExists` method, but just do this:
```typescript
const key = this.toKey({factoryId, options});
return !!this.doGetWidget(key);
```
Or, alternatively, change the `getWidget()` code slightly:
```typescript
    getWidget<T extends Widget>(factoryId: string, options?: any): MaybePromise<T> | undefined {
        const key = this.toKey({ factoryId, options });
        return this.doGetWidget<T>(key);
    }
```
By making `getWidget` not async, so that it doesn't *necessarily* return a promise, you get the same ability to check for the existence of a widget from outside the widget manager by checking !!widgetManager.getWidget(...)`

File: packages/core/src/browser/widget-open-handler.ts [hunk 7]
```diff
@@ -86,16 +86,17 @@ export abstract class WidgetOpenHandler<W extends BaseWidget> implements OpenHan
...
     async open(uri: URI, options?: WidgetOpenerOptions): Promise<W> {
-        const widget = await this.getOrCreateWidget(uri, options);
-        await this.doOpen(widget, options);
+        const widgetExists = this.widgetExists(uri, options);
+        const widget = widgetExists ? (await this.getWidget(uri, options))! : await this.createWidget(uri, options);
+        await this.doOpen(widget, widgetExists, options);
         return widget;
     }
-    protected async doOpen(widget: W, options?: WidgetOpenerOptions): Promise<void> {
+    protected async doOpen(widget: W, widgetExists: boolean, options?: WidgetOpenerOptions): Promise<void> {
```
Comment (colin-grant-work) [inline packages/core/src/browser/widget-open-handler.ts:94]: You could avoid having to change the signature of this method by adding an optional field to the `WidgetOpenerOptions` interface (or making a new one, something like `WidgetDoOpenerOptions`) that has a `widgetAlreadyCreated` field and adding the `widgetExists` information that way.

Comment (kena0ki) [reply packages/core/src/browser/widget-open-handler.ts:94]: Thank you for the review.
Do you mean we should not delete `doOpen` override method in `git-commit-detail-open-handler.ts` or we just want to reduce the number of parameters?
I prefer the latter suggestion, i.e. creating `WidgetDoOpenerOptions`, since the `widgetExists` information is not necessary for `open` method. 
However, If we shouldn't delete `doOpen` method, the latter suggestion can't be chosen, since the signature of `doOpen` method in `git-commit-detail-open-handler.ts` and in `widget-open-handler.ts` would be vary.

Comment (colin-grant-work) [reply packages/core/src/browser/widget-open-handler.ts:94]: > Thank you for the review.
> Do you mean we should not delete `doOpen` override method in `git-commit-detail-open-handler.ts` or we just want to reduce the number of parameters?

Yes, just reducing the parameters so that this conforms more closely to the signature of `doOpen` in `WidgetOpenHandler` sounds good to me. That means it shouldn't break any downstream functionality - they'll just get the behavior without setting `widgetExists`.

**Comment by kena0ki** [comment]
I've updated the PR but I found that `getOrCreateWidget` method in sub classes, `editor-manager.ts` and `mini-browser-open-handler.ts`, needs to be changed. It's because keys (options) to get widgets are different from parent class `widget-open-handler.ts`.

I'll close this PR since I unable to reproduce the issue #8511 now on master branch.
